ci: modify bpftrace script to ignore destinations outside pod CIDRs

smagnani96 opened the pull request:
With this PR, the bpftrace script that we run in CI is now able to ignore TCP traffic with destionation address outside pod CIDRs. This is particularly useful in egress-gateway tests, for which pod-to-world and pod-to-node traffic is sent.
Prior to this, in conformance-ipsec-e2e we used to keep tests separate, and run bpftrace only on non egress-gateway tests.
This PR (re)unifies tests while also running bpftrace in background for all of them.

```release-note
Modify bpftrace script in CI to ignore proxy traffic if destination is outside pod CIDRs.
```

smagnani96 committed:
ci: bpftrace skip tracking connections outside pod CIDRs

The check-ipsec-leak.bt script reports errors for plain-text packets
(non-ESP only, this is not used for Wireguard) captured from bridge when:

1. the packets have pod IP as both src and dst
2. or the packets have either src or dst IP inside pod CIDR, plus the
   packets are from proxy ($trace_override)

Let's focus on 2.

In the normal case of pod->world via L7, proxy emits packets like pod->world,
but masquerading (nf/bpf) snats the packets to node->world, therefore
the script never sees pod IP on bridge, no errors are reported.

In the case of egress gateway via L7, the packet is first forwarded to the
gateway node after L7 policy check with the pod->world notation, and the
gateway node later forwards the request using to gateway->world.
Here the script sees that 1st packet with src pod IP && from_proxy,
triggering the (false) alarm.

This PR is to make sure that these type of packets won't be tracked as
from_proxy ($trace_override). We currently avoid this "false" alarm by
running egress-gateway connectivity tests separately from all the others
and with no leak-detection script running.
As we're going to introduce leak detection in more and more CI tests
hopefully, let's add this check to skip tracking unwanted traffic.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
conformance-ipsec-e2e: do not split egress-gateway tests

In the previous commit, we introduced to the bpftrace the capability to
skip tracking traffic with destination address outside pod CIDRs.
At this point, in conformance-ipsec-e2e tests we do not need to split tests
anymore (egress-gateway vs all the others), previously modified in #35323,
"workflows/ipsec: Cover egress gateway", ea6bdac11eae9ceb0dd9524cd42f8cf61a9f276b.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

pchaigno commented on .github/actions/bpftrace/scripts/check-ipsec-leaks.bt:
```diff
@@ -165,6 +165,7 @@ kprobe:br_forward
 }
 
```
> This might cause problems in egress-gateway tests, since pod-to-outside and pod-to-node traffic is generated.

Pod-to-outside and pod-to-node traffic is also generated without egress gateway enabled. Did you mean something more specific, like this traffic is generated on the overlay when egress gateway is enabled?

smagnani96 replied:
What I was referring to (apologize, it was not that clear) is when having the L7 proxy around. Let me clarify, iiuc.

This portion of the script tracks only TCP connections established by the L7 proxy:
https://github.com/cilium/cilium/blob/6a4a1d17271a34851ff02c662379cb197bd95f08/.github/actions/bpftrace/scripts/check-ipsec-leaks.bt#L169-L173

Then in the `kprobe:br_forward` (interface that connects all nodes, where we must see only encrypted packets for pod-to-pod communication), we check all the non-ESP packets:
https://github.com/cilium/cilium/blob/6a4a1d17271a34851ff02c662379cb197bd95f08/.github/actions/bpftrace/scripts/check-ipsec-leaks.bt#L130

For those plain-text packets, we do check if the connection is traced:
https://github.com/cilium/cilium/blob/6a4a1d17271a34851ff02c662379cb197bd95f08/.github/actions/bpftrace/scripts/check-ipsec-leaks.bt#L134-L136

In case either the source or destination address is a pod, we then log the packet (causing the test to fail, since output is not empty):
https://github.com/cilium/cilium/blob/6a4a1d17271a34851ff02c662379cb197bd95f08/.github/actions/bpftrace/scripts/check-ipsec-leaks.bt#L138-L140

Taking as an example the following actions from the `seq-egress-gateway-with-l7-policy` tests:

```bash
.  [.] Action [seq-egress-gateway-with-l7-policy/egress-gateway/curl-external-echo-service-2: cilium-test-1/client3-75555c5f5-h27hq (10.244.3.231) -> cilium-test-1/echo-external-node (echo-external-node.cilium-test-1:8080)]
.  [.] Action [seq-egress-gateway-with-l7-policy/egress-gateway/curl-external-echo-pod-0: cilium-test-1/client3-75555c5f5-h27hq (10.244.3.231) -> cilium-test-1/echo-external-node-55984f6c4f-b9l6d (172.18.0.5:8080)]

```

in this case:

* the destination is an external node
* source is a pod in a Cilium-managed node

In the script, the packet would match:
* non-ESP
* `trace_override=1` (without this patch)
* `$src_is_pod`, as source address doesn't get masqueraded

Resulting in the bpftrace script logging the two `curl`s following:

```bash
[16:42:26:600207] 10.244.3.231:52706 -> 172.18.0.5:8080 (proto: 6, TCP flags: S..., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:600335] 172.18.0.5:8080 -> 10.244.3.231:52706 (proto: 6, TCP flags: SA.., encap: 1, ifindex: 51, netns: f0000000, override: 1)
[16:42:26:600396] 10.244.3.231:52706 -> 172.18.0.5:8080 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:600524] 10.244.3.231:52706 -> 172.18.0.5:8080 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:600562] 172.18.0.5:8080 -> 10.244.3.231:52706 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 51, netns: f0000000, override: 1)
[16:42:26:601558] 172.18.0.5:8080 -> 10.244.3.231:52706 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 51, netns: f0000000, override: 1)
[16:42:26:601628] 10.244.3.231:52706 -> 172.18.0.5:8080 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:653119] 10.244.3.231:52714 -> 172.18.0.5:8080 (proto: 6, TCP flags: S..., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:653291] 172.18.0.5:8080 -> 10.244.3.231:52714 (proto: 6, TCP flags: SA.., encap: 1, ifindex: 51, netns: f0000000, override: 1)
[16:42:26:653365] 10.244.3.231:52714 -> 172.18.0.5:8080 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:653527] 10.244.3.231:52714 -> 172.18.0.5:8080 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 49, netns: f0000000, override: 1)
[16:42:26:653585] 172.18.0.5:8080 -> 10.244.3.231:52714 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 51, netns: f0000000, override: 1)
[16:42:26:654547] 172.18.0.5:8080 -> 10.244.3.231:52714 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 51, netns: f0000000, override: 1)
[16:42:26:654618] 10.244.3.231:52714 -> 172.18.0.5:8080 (proto: 6, TCP flags: .A.., encap: 1, ifindex: 49, netns: f0000000, override: 1)
```

jschwinger233 replied:
(`$trace_override` can have a better name `$from_proxy`, or more specific after this PR: `$pod_to_pod_via_proxy`)

Check-ipsec-leak.bt script is reporting errors for plain-text packets captured from bridge, as long as:
1. the packets have pod IP as both src and dst
2. or the packets have either src or dst IP inside pod CIDR, plus the packets are from proxy ($trace_override)

Let's focus on 2.

In the normal case of pod->world via L7, proxy emits packets like `pod->world`, but masquerading (nf / bpf) snats the packets to `node->world`, therefore check-ipsec-leak.bt never saw pod IP on bridge, no errors were reported.

In the case of egress gateway via L7, I believe the packets must skip the masquerading, then check-ipsec-leak.bt sees packets with src pod IP && from_proxy, triggers the (false) alarm. 

This PR is to make sure those traffic won't be marked as `from_proxy` ($trace_override).

smagnani96 replied:
I followed your suggestion and added a commit to change the `$trace_override` name to `$pod_to_pod_via_proxy` to make things a bit clearer.

pchaigno replied:
> In the case of egress gateway via L7, I believe the packets must skip the masquerading

And that's because we're talking about the packet as they are leaving the source node, after they went through L7 egress policies, and before they reach the gateway node?

And this isn't an issue for WireGuard because, with WireGuard, egress gateway traffic between cluster nodes is encrypted? Or is this script simply not used for WireGuard?

smagnani96 replied:
> And that's because we're talking about the packet as they are leaving the source node, after they went through L7 egress policies, and before they reach the gateway node?

I think yes, this is a correct case: `pod->world` as for `pod->external-pod`  (eg., in the egw tests) would lead to always a plain-text packet, which is first forwarded from node to egress gateway node with `pod_ip -> external-pod-ip` **triggering the tracing as `$from_proxy` and `$src_pod`** , then to the external endpoint with `egress_ip -> external-pod-ip`.

> And this isn't an issue for WireGuard because, with WireGuard, egress gateway traffic between cluster nodes is encrypted? Or is this script simply not used for WireGuard?

The script is being used only in conformance-ipsec-e2e (and soon in conformance-ipsec-upgrade).

pchaigno replied:
Thanks for all the explanation!

To go back to my original comment, could you add a bit of the context (I think Gray's explanation is at the right level) to the commit description?

smagnani96 replied:
Thanks both of you for the feedback.
I just changed the 1st commit message description with the discussed change :+1:

pchaigno reviewed (commented):
Nice!

One question below. I'm not very familiar with the bpftrace script, so trying to make sure I understand things correctly.

smagnani96 committed:
ci: ipsec: rename trace_override variable in bpftrace script

This commit renames the `$trace_override` variable in the bpftrace script
to `$pod_to_pod_via_proxy`. We use the variable to identify packets that
are generated from the L7 proxy and with destination an IP within the
pod CIDRs, so let's clarify the name of the variable accordingly.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

pchaigno reviewed (approved):
Looks good to me. Only thing is my comment above on the first commit's description.
